libraries/botbuilder/src/fileTranscriptLogger.ts

```diff
@@ -0,0 +1,81 @@
+/**
+ * @module botbuilder
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+// import * as fs from 'async-file';
+import * as fsx from 'async-file';
+import * as filenamify from 'filenamify';
+import * as fs from 'fs';
+import * as path from 'path';
+import { Activity, TranscriptLogger } from '.';
+
+/**
+ * The file transcript logger creates .transcript file for each conversation it logs
+ *
+ * @remarks
+ * This class logs all activities for a conversation to a {conversationId}.transcript file 
+ * This is useful to create .transcript files for unit tests with the TestAdapter
+ *
+ * Below is the boilerplate code needed to use this in your app:
+ * ```javascript
+ * const { FileTranscriptLogger, TranscriptLoggerMiddleware } = require('botbuilder');
+ *
+ * adapter.use(new TranscriptLoggerMiddleware(new FileTranscriptLogger()));
+ * ```
+ */
+export class FileTranscriptLogger implements TranscriptLogger {
+
+    private folder: string;
+    private started = new Set<string>();
+
+    /**
+     * Creates an instance of FileTranscriptLogger.
+     * @param folder Root folder where transcript will be stored. default is __dirname
+     */
+    constructor(folder: string = null, private unitTestMode: boolean = true) {
+        this.folder = folder || process.cwd();
+        
+        if (!fs.existsSync(this.folder)) {
+            fs.mkdirSync(this.folder);
+        }
+    }
+
+    /**
+     * Log an activity to the transcript.
+     * @param activity Activity being logged.
+     */
+    public async logActivity(activity: Activity): Promise<void> {
+        if (!activity) {
+            throw new Error('activity cannot be null for logActivity()');
+        }
+
+        const transcriptPath: string = this.getTranscriptPath(activity.conversation.id);
+        let activities: Activity[] = [];
+        if (await fsx.exists(transcriptPath)) {
+            if (this.unitTestMode === true && this.started.has(transcriptPath) === false) {
+                this.started.add(transcriptPath);
+            }
+            else {
+                const json = await fsx.readFile(transcriptPath, 'utf8');
```

Comment:
This code has the same bug as the C#. Over time, performance will progressively degrade as the file gets larger and larger. 

This code also has bugs in that multiple processes / activities will lead to data loss on the file. 

The design of the file format needs to be revisited.

Reply:
renamed to unitTestTranscriptLogger to clearly delineate this is not for production code.  FileTranscriptStore correctly stores activities in performant way but which is not easy to consume by loading into emulator directly.